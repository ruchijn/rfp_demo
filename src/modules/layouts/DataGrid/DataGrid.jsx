import React from 'react';
import PropTypes from 'prop-types';
import isEmptyObj from '../../../utils/objectUtil';

import {
  DataGridContainer,
  FetchDataButtonWrapper,
  FetchDataButton,
  TextWrapper,
  Text,
  TopInformationWrapper,
  TopInformationSection,
  RatesWrapper,
  RatesRow,
  RatesCol,
} from './styledComponents';

class DataGrid extends React.Component {
  handleButtonClick = e => {
    const { fetchRates } = this.props;
    e.preventDefault();
    fetchRates();
  };

  renderText = (displayText, fontColor = 'regular') => {
    return (
      <TextWrapper>
        <Text color={fontColor}>{displayText}</Text>
      </TextWrapper>
    );
  };

  renderConversionRow = rateEntry => {
    const currency = rateEntry[0];
    const value = rateEntry[1];

    return (
      <RatesRow key={currency}>
        <RatesCol rightBorder>{currency}</RatesCol>
        <RatesCol>{value}</RatesCol>
      </RatesRow>
    );
  };

  renderData = () => {
    const {
      rates: { base, date, rates },
    } = this.props;

    const baseCurrency = this.renderText(`Base currency: ${base}`);
    const conversionDate = this.renderText(`Conversion date: ${date}`);
    const rateEntries = !isEmptyObj(rates) ? Object.entries(rates) : [];

    return (
      <React.Fragment>
        <TopInformationWrapper>
          <TopInformationSection>{baseCurrency}</TopInformationSection>
          <TopInformationSection right>{conversionDate}</TopInformationSection>
        </TopInformationWrapper>
        <RatesWrapper>
          {rateEntries && rateEntries.map(rateEntry => {
            return this.renderConversionRow(rateEntry);
          })}
        </RatesWrapper>
      </React.Fragment>
    );
  };

  renderError = () => {
    const displayText = 'We are sorry. We are not able to get the data you requested.';
    return this.renderText(displayText, 'red');
  };

  renderLoader = () => {
    return this.renderText('Loading...', 'shuttleGrey');
  };

  renderSection = () => {
    const { isLoading, error, rates } = this.props;
    if (isLoading) return this.renderLoader();

    if (!isEmptyObj(error)) return this.renderError();

    if (!isEmptyObj(rates)) return this.renderData();

    return this.renderText('Please click on button to get the conversion rates');
  };

  render = () => {
    return (
      <DataGridContainer>
        <FetchDataButtonWrapper>
          <FetchDataButton onClick={e => this.handleButtonClick(e)}>
            Click me to get data
          </FetchDataButton>
        </FetchDataButtonWrapper>
        {this.renderSection()}
      </DataGridContainer>
    );
  };
}

DataGrid.propTypes = {
  fetchRates: PropTypes.func.isRequired,
  isLoading: PropTypes.bool,
  error: PropTypes.object,
  rates: PropTypes.object,
};

DataGrid.defaultProps = {
  isLoading: false,
  error: {},
  rates: {},
};

export default DataGrid;
