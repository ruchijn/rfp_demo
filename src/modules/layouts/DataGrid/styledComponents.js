import styled from 'styled-components';
import { colorLightGrey, colorShuttleGrey, colorCardinal } from '../../../colorVariables';

export const DataGridContainer = styled.div({
  display: 'flex',
  maxWidth: '80%',
  justifyContent: 'center',
  margin: '5rem auto',
  flexDirection: 'column',
});

export const FetchDataButtonWrapper = styled.div({
  width: '100%',
  display: 'flex',
  justifyContent: 'flex-end',
  marginBottom: '1rem',
});

export const FetchDataButton = styled.button({
  border: `1px solid ${colorLightGrey}`,
  padding: '1rem 2rem',
});

export const TextWrapper = styled.div({
  marginTop: '1rem',
});

export const Text = styled.p(({ color }) => ({
  fontSize: '1.5rem',
  textAlign: 'center',
  ...(color === 'shuttleGrey' && {
    color: colorShuttleGrey,
  }),
  ...(color === 'red' && {
    color: colorCardinal,
  }),
}));

export const TopInformationWrapper = styled.div({
  display: 'flex',
  flexDirection: 'column',
  '@media screen and (min-width: 992px) {': {
    flexDirection: 'row',
  },
});

export const TopInformationSection = styled.div(({ right }) => ({
  display: 'flex',
  '@media screen and (min-width: 992px) {': {
    flex: '1 1 50%',
    ...(right && {
      justifyContent: 'flex-end',
    }),
  },
}));

export const RatesWrapper = styled.div({
  display: 'flex',
  flexDirection: 'column',
  border: `1px solid ${colorLightGrey}`,
  '&:last-child {': {
    borderBottom: 'none',
  },
});

export const RatesRow = styled.div({
  display: 'flex',
  flexDirection: 'row',
  borderBottom: `1px solid ${colorLightGrey}`,
});

export const RatesCol = styled.div(({ rightBorder }) => ({
  flex: '1 1 50%',
  padding: '0.5rem 1rem',
  ...(rightBorder && {
    borderRight: `1px solid ${colorLightGrey}`,
  }),
}));
