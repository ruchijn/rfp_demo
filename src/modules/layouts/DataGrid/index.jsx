import { connect } from 'react-redux';
import { fetchRates } from '../../../actions/conversionActions';

import DataGrid from './DataGrid';

export const mapStateToProps = state => {
  const { rates, error, isLoading } = state.conversion;

  return {
    rates,
    error,
    isLoading,
  };
};

export const mapDispatchToProps = dispatch => ({
  fetchRates: () => {
    dispatch(fetchRates());
  },
});

const getConnectedDataGrid = connect(
  mapStateToProps,
  mapDispatchToProps,
)(DataGrid);

export default getConnectedDataGrid;
