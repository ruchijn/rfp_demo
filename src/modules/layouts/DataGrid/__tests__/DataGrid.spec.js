import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import 'jest-styled-components';
import DataGrid from '../DataGrid';

describe('<DataGrid />', () => {
  const fetchRates = jest.fn();

  const defaultProp = {
    fetchRates,
  };

  const e = { preventDefault: jest.fn() };
  let dataGridWrapper;

  beforeEach(() => {
    dataGridWrapper = shallow(<DataGrid {...defaultProp} />);
  });

  it('matches the snapshot', () => {
    expect(shallowToJson(dataGridWrapper)).toMatchSnapshot();
  });

  it('renders loading while fetching data', () => {
    const loadingProp = {
      ...defaultProp,
      isLoading: true,
    };
    const loadingWrapper = shallow(<DataGrid {...loadingProp} />);
    expect(loadingWrapper.find('styledComponents__Text').props().color).toEqual('shuttleGrey');
  });

  it('renders error when error passed', () => {
    const errorProp = {
      ...defaultProp,
      error: {
        status: '422',
        message: 'Error occurred',
      },
    };
    const errorWrapper = shallow(<DataGrid {...errorProp} />);
    expect(errorWrapper.find('styledComponents__Text').props().color).toEqual('red');
  });

  it('renders data when data returned', () => {
    const dataProp = {
      ...defaultProp,
      rates: {
        base: 'CAD',
        date: '2019-09-20',
        rates: {
          USD: '23.98',
          EUR: '12.23',
        },
      },
    };
    const dataWrapper = shallow(<DataGrid {...dataProp} />);
    expect(dataWrapper.find('styledComponents__TopInformationWrapper')).toHaveLength(1);
  });

  it('renders data when no rates returned', () => {
    const dataProp = {
      ...defaultProp,
      rates: {
        base: 'CAD',
        date: '2019-09-20',
      },
    };
    const dataWrapper = shallow(<DataGrid {...dataProp} />);
    expect(dataWrapper.find('styledComponents__TopInformationWrapper')).toHaveLength(1);
  });

  describe('test static functions', () => {
    describe('handleButtonClick', () => {
      it('calls handleButtonClick', () => {
        dataGridWrapper.instance().handleButtonClick(e);

        expect(fetchRates).toBeCalled();
      });
    });
  });
});
