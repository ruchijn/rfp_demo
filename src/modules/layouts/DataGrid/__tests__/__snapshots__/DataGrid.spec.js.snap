// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DataGrid /> matches the snapshot 1`] = `
<styledComponents__DataGridContainer>
  <styledComponents__FetchDataButtonWrapper>
    <styledComponents__FetchDataButton
      onClick={[Function]}
    >
      Click me to get data
    </styledComponents__FetchDataButton>
  </styledComponents__FetchDataButtonWrapper>
  <styledComponents__TextWrapper>
    <styledComponents__Text
      color="regular"
    >
      Please click on button to get the conversion rates
    </styledComponents__Text>
  </styledComponents__TextWrapper>
</styledComponents__DataGridContainer>
`;
